src/backend/cdb/cdbmutate.c

```diff
@@ -366,6 +376,48 @@ get_partitioned_policy_from_flow(Plan *plan)
 									   GP_POLICY_DEFAULT_NUMSEGMENTS());
 }
 
+static bool
+broadcast_motion_walker(Node *node, RowsMutator *state)
+{
+	if (node == NULL)
+		return false;
+
+	if (is_plan_node(node) &&
+		((Plan*)node)->flow != NULL &&
+		((Plan*)node)->flow->req_move == MOVEMENT_BROADCAST)
+	{
+		state->scaleFactor = ((Plan*)node)->flow->numsegments;
+		plan_tree_walker(state->sliceRoot, rows_number_walker, state);
+	}
+	else if (IsA(node, Motion) || IsA(node, SubPlan))
+	{
+		state->sliceRoot = node;
+	}
+
+	return plan_tree_walker(node, broadcast_motion_walker, state);
+}
+
+static bool
+rows_number_walker(Node *node, RowsMutator *state)
+{
+	Plan *currentPlan;
+
+	if (node == NULL)
+		return false;
+
+	if (is_plan_node(node) &&
+		((Plan*)node)->flow != NULL &&
+		((Plan*)node)->flow->req_move == MOVEMENT_BROADCAST)
+		return true;
+
+	if (is_plan_node(node))
+	{
+		currentPlan = (Plan*)node;
+		currentPlan->plan_rows *= state->scaleFactor;
+	}
```

Comment:
```suggestion
	if (is_plan_node(node))
	{
		if (((Plan*)node)->flow != NULL &&
			((Plan*)node)->flow->req_move == MOVEMENT_BROADCAST)
			return true;

		currentPlan = (Plan*)node;
		currentPlan->plan_rows *= state->scaleFactor;
	}
```

Maybe combine if statements?

Reply:
Applied